sycl/doc/SYCL_compiler_and_runtime_design.md

```diff
@@ -0,0 +1,377 @@
+- the offload wrapper object file
+- all the host libraries needed by the application
+- the SYCL runtime library
+
+The result is so-called "fat binary image" containing the host code, code for
+all the targets plus the registration/unregistration functions and the
+information about the device binary images.
+
+When compilation and linking is done in single compiler driver invocation, the
+bundling and unbundling steps are skipped.
+
+*Design note: the described scheme differs from current llvm.org
+implementation.  Current design uses Linux-specific linker script approach and
+requires that all the linked fat objects are compiled for the same set of
+targets. The described design uses OS-neutral offload-wrapper tool and does not
+impose restrictions on fat objects.*
+
+### Integration with SPIR-V format
+
+This section explains how to generate SPIR-V specific types and operations from
+C++ classes and functions.
+
+Translation of SYCL C++ programs to the code executable on heterogeneous
+systems can be considered as three step process:
+1) translation of SYCL C++ programs into LLVM IR
+1) translation from LLVM IR to SPIR-V
+1) translation from SPIR-V to machine code
+
+LLVM-IR to SPIR-V translation is performed by a dedicated tool -
+[translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator).
+This tool correctly translates most of regular LLVM IR types/operations/etc to
+SPIR-V.
+
+For example:
+
+- Type: `i32` --> `OpTypeInt`
+- Operation: `load` --> `OpLoad`
+- Calls: `call` --> `OpFunctionCall`
+
+SPIR-V defines special built-in types and operations that do not have
+corresponding equivalents in LLVM IR. E.g.
+
+- Type: ??? --> `OpTypeEvent`
+- Operation: ??? --> `OpGroupAsyncCopy`
+
+Translation from LLVM IR to SPIR-V for special types is also supported, but
+such LLVM IR must comply to some special requirements. Unfortunately there is
+no canonical form of special built-in types and operations in LLVM IR, moreover
+we can't re-use existing representation generated by OpenCL C front-end
+compiler.  For instance here is how OpGroupAsyncCopy operation looks in LLVM IR
+produced by OpenCL C front-end compiler.
+
+```LLVM
+@_Z21async_work_group_copyPU3AS3fPU3AS1Kfjj(float addrspace(3)*, float addrspace(1)*, i32, i32)
+```
+
+It's a regular function, which can conflict with user code produced from C++
+source.
+
+
+SYCL compiler uses solution developed for OpenCL C++ compiler prototype:
+
+- Compiler: https://github.com/KhronosGroup/SPIR/tree/spirv-1.1
+- Headers: https://github.com/KhronosGroup/libclcxx
+
+SPIR-V types and operations that do not have LLVM equivalents are **declared**
+(but not defined) in the headers and satisfy following requirements:
+
+- the type must be pre-declared as a C++ class in `cl::__spirv` namespace
+- the type must not have actual definition in C++ program
+- the operation is expressed in C++ as `extern` function not throwing C++
+  exceptions
```

Comment:
I have the feeling that in sycl/include/CL/__spirv/spirv_ops.hpp there are still a lot of `noexcept` missing...

Reply:
I guess this comment doesn't imply changes in the documentation, but rather in the source code. Right?

Reply:
Actually I do not remember what I wanted to say so long ago...